src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/IncludeDenormalizedRewriter.cs

```diff
@@ -0,0 +1,44 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Health.Fhir.Core.Features.Search.Expressions;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Search.Expressions.Visitors
+{
+    internal class IncludeDenormalizedRewriter : SqlExpressionRewriterWithInitialContext<object>
+    {
+        public static readonly IncludeDenormalizedRewriter Instance = new IncludeDenormalizedRewriter();
+
+        public override Expression VisitSqlRoot(SqlRootExpression expression, object context)
+        {
+            if (expression.TableExpressions.Count == 0)
+            {
+                return expression;
+            }
+
+            // If the only table expressions are includes, then we need to hoist denormalized expressions to the top.
+            bool onlyIncludes = expression.TableExpressions.All(te => te.Kind == TableExpressionKind.Include);
+            if (onlyIncludes)
+            {
+                var newNormalizedPredicates = new List<TableExpression>(expression.TableExpressions.Count + 1);
+                newNormalizedPredicates.AddRange(expression.TableExpressions);
+
+                Expression denormalizedExpression = expression.DenormalizedExpressions.Count > 1 ?
+                    new MultiaryExpression(MultiaryOperator.And, expression.DenormalizedExpressions)
+                    : expression.DenormalizedExpressions[0];
```

Comment:
Count could be 0?

Reply:
Maybe? I need to dig into the spec and see if an `_include` parameter is valid on a system-wide search. If not, there should always be at least one denormalized expression of the original resource type.

Reply:
It looks like I can get away with just doing `new MultiaryExpression(MultiaryOperator.And, expression.DenormalizedExpressions)` in all cases.